### PR TITLE
Added relation provider compiler pass

### LINKED ISSUE #27: Extension point (tag) to register a relation provider resolver
See:
- https://github.com/willdurand/hateoas#relation-provider
- #15

### PR DESCRIPTION
**Pull request by dantleech**
This PR adds a compiler pass which adds services tagged with `hateoas.relation_provider` to the chain provider.

Aims to fix #27

### CODE REVIEW

File: Tests/DependencyInjection/BazingaHateoasExtensionTest.php [hunk 1]
```diff
@@ -50,6 +51,16 @@ public function testLoad()
         );
     }
 
+    public function testRelationProviderPass()
+    {
+        $container = $this->getContainerForConfig(array(array()));
+        $container->addCompilerPass(new RelationProviderPass());
```
Comment (dantleech) [inline Tests/DependencyInjection/BazingaHateoasExtensionTest.php]: Would have preferred to put this in `Tests/DependencyInjection/Compiler/RelationProviderTest` but would then need to create a new abstract `TestCase` class with the clean up methods. Can do this if required.

**Comment by dantleech** [comment]
Hmm. this actually only adds the `resolver` services which is not exactly what I intended.

- Do we still want to add tagged resolvers?
- Shall I add/replace a/the compiler pass for adding `RelationProvider` instances?

File: DependencyInjection/Compiler/RelationProviderPass.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        $registryDefinition = $container->getDefinition('hateoas.configuration.provider.resolver.chain');
+
+        $relationProviderDefinitions = array();
+        foreach (array_keys($container->findTaggedServiceIds('hateoas.relation_provider')) as $id) {
+            $definition = $container->getDefinition($id);
+            $this->assertProvider($container, $definition);
+            $relationProviderDefinitions[] = $definition;
+        }
+
+        $registryDefinition->setArguments($relationProviderDefinitions);
```
Comment (adrienbrault) [inline DependencyInjection/Compiler/RelationProviderPass.php]: This is wrong, it should be `->setArguments([$relationProviderDefinitions]);` or `->replaceArgument(0, $relationProviderDefinitions);`

**Comment by adrienbrault** [comment]
@dantleech You cannot add a RelationProvider, you could only replace it, which would "break" the library. Adding a RelationProviderResolverInterface is what you should want.

:+1: for the code once my comment is addressed

**Comment by dantleech** [comment]
Updated.

**Comment by adrienbrault** [comment]
Looks good :+1:

**Comment by willdurand** [comment]
Cool! Can I ask for some documentation? :fearful:

File: Resources/doc/index.md [hunk 3]
```diff
@@ -69,6 +63,34 @@ class AppKernel extends Kernel
...
+are no special considerations:
+
+````php
+// My/Controller.php
+
+class SomeController extends Controller
+{
+    public function resourceAction(Request $request)
+    {
+        $post = $repository->find('BlogBundle:post');
```
Comment (dantleech) [inline Resources/doc/index.md]: Added this section because it wasn't immediately clear to me how to use this bundle :)

**Comment by dantleech** [comment]
Added some documentation, would have added some more detail but I havn't actually tried this feature out yet.

I am trying to get Hateoas support in the [SonataAdminBundle](https://github.com/sonata-project/SonataAdminBundle/pull/2684) so that it can add links for the Symfony CMF [ResourceRestBundle](https://github.com/symfony-cmf/ResourceRestBundle).

**Comment by willdurand** [comment]
Thanks

**Comment by kayue** [comment]
@dantleech @willdurand It seems the chain resolver isn't being applied automatically. I have to do the following in order to use the chain resolver.

```
Product:
    exclusion_policy: ALL
    relation_providers: [hateoas.configuration.provider.resolver.chain]
```

**Pull request merged**